### DISCUSSION TITLE
Should haves

### DISCUSSION THREAD

**Discussion by jhpratt** [rust-marker/design#2] [Ideas]
Use this discussion for suggestions of "should haves" for a custom linter. Please limit suggestions to one per top-level comment to allow for easier upvoting/downvoting and replies.

**Comment by jhpratt** [comment]
Lint authors should only need to use the stable compiler.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
Lints should be easy to write, install, and use.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
Code suggestions should be able to be provided and (possibly) automatically applied.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
A command to validate that a lint implementation upholds the expected API should be provided.

**Reply by flip1995** [reply]
I'm not sure what this means, could you clarify?

**Reply by jhpratt** [reply]
`foo validate bar.wasm` could fail if `bar.wasm` doesn't actually implement the linting API expected by the linter. Mainly useful for CI purposes, although a registry could also require it pass to upload in theory.

**Reply by flip1995** [reply]
I still don't understand. I think we have to clarify 2 things first: The nomenclature and if we really should use wasm.

**Reply by jhpratt** [reply]
Because a lint implementation wouldn't be statically linked with the linting tool, it would be a runtime error if the implementation doesn't implement the expected API. While hopefully unlikely, it seems plausible that someone writing a lint for the first time could get the public API wrong. Providing a command to make sure they didn't do that would help them out.

Now that I think about it, this is probably a "could have".

I'm signing off for a bit, but I'll certainly continue conversation at some point. Might be worthwhile to set up a thread on Zulip and/or have a call at some point to discuss the major points.

**Reply by flip1995** [reply]
Shouldn't we design this API in a way, that when using it, you can be sure that your lints will work?

Agree to set up a call to talk about it.

**Reply by jhpratt** [reply]
If the linting tool requires that all lint implementations have `fn foo(_: Cat)`, but the lint implementation actually has `fn foo(_: Dog)`, I'd like to be able to catch that. As I said it is already doable and would be a runtime error, but why not provide a command that would ensure it has the right API before publishing? Or (depending on #6) this could be done by the registry and refuse publication if it doesn't implement the right API.

This command/validation would be for lint authors, not end users. An end user should assume that lints "just work".

**Reply by flip1995** [reply]
Why not dogfood this and write a custom lint with the stable API for usage of the stable API? :)

**Reply by jhpratt** [reply]
That's actually not a bad idea haha

**Reply by flip1995** [reply]
We perfected this approach in Clippy. Since yesterday, we use a lint to generate lint documentation :smile:

**Reply by jhpratt** [reply]
I'm going to go ahead and withdraw this idea, as having a lint of our own that ensures the only public items are the lints that follow the expected API seems more than reasonable.

**Comment by jhpratt** [comment]
Authoring a lint should be easy.

**Reply by flip1995** [reply]
This is a bit of a subjective statement. But I generally agree with it.

**Reply by jhpratt** [reply]
Certainly subjective. Just something to keep in mind imo.

**Comment by DevinR528** [comment]
Crate-specific lints should be able to be bundled with the library they are for and read/registered by the driver.

**Reply by jhpratt** [reply]
I've thought about this. Bundling with the library seems infeasible unless we auto-detect direct dependencies. Even then, we'd have to validate that the lint author can publish that crate on crates.io (which requires the full crate registry) _and_ restrict publishing to those people.

**Reply by DevinR528** [reply]
Is this way to hacky?
```rust
// inside a crate with exports that should be linted
#[cfg(feature = "linter_plugin")] // the driver would need to be able to turn this on
#[no_mangle]
pub static PLUGIN_BYTES: &[u8] = include_bytes!("plugin.wasm"); // a wasm compiled lint
```
then load the dylib using `libloading` (yeah this is getting wonky) and get the bytes and load the wasm.

**Reply by jhpratt** [reply]
Something about that approach just _feels_ wrong. I think having a separate registry,  even if just at first, makes more sense. I presume some company would be willing to sponsor the hosting, but that's speculation of course. An advantage there is that the registry could optimize the wasm upon upload.

**Reply by jhpratt** [reply]
I think this would be best as a _possible_ long-term goal that would go alongside rustc/cargo/crates.io integration, if that ultimately happens.

**Comment by jhpratt** [comment]
Lint groups should exist and be trivial to create.

**Reply by jhpratt** [reply]
As an addendum, having a default "all" group for each crate would be convenient.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
It should be possible to deprecate a lint.

**Reply by jhpratt** [reply]
As no one objected, I have added this to the list of agreed upon items.

**Comment by jhpratt** [comment]
Lints should be versioned.

**Reply by flip1995** [reply]
I would make this a "Could" requirement. This is a concept, that doesn't exist in Clippy or Rust and we never missed it.

**Reply by jhpratt** [reply]
I think it is at least relatively important because this would be tracking something that could have breaking changes (that being third-party crates), whereas clippy and rustc are basically stable (I know clippy has some for regex, which could in theory issue a breaking release and break the lints).

**Reply by flip1995** [reply]
Ah I see. But I see the custom lints that crates implement are just another crate using this API. So it can be versioned just as any other crate. So I don't think lints have to be versioned separately. This just introduces confusion for the users of the custom linter.

**Reply by jhpratt** [reply]
Let's resolve this as using whatever versioning the registry provides? Whether that is crates.io or otherwise.

**Reply by jhpratt** [reply]
Resolving as such.

**Comment by xFrednet** [comment]
Support features for lints like they already work for dependencies:
* This means supporting optional lints, similar to how the `optional = 1` flag works for dependencies 
* Lint code should support conditional compilation and optional dependencies

**Reply by jplatte** [reply]
> Lint code should support conditional compilation and optional dependencies

Why? This pretty much precludes distributing lints in wasm form (only), which seems like a nice thing to have (eventually).

**Reply by xFrednet** [reply]
It should still be possible to write code that checks for platform/feature specific pitfalls in a wasm context.

I imagine this to be used for crates with optional features, then you might emit a different message depending on the enabled features. This might not be the most important usage, hence it being in *should have*, but it would definitely be nice. Also, because natively, Rust supports the usage of `cfg` and it might confuse users if this is not really usable in lint code.

**Reply by jplatte** [reply]
Wait, are you talking about the lint implementation itself using conditional compilation (so you could compile it with different feature sets before it ever runs) or about the lint being able to inspect cfg attributes in the codebase being linted? I agree emitting different messages from `foobar-lint` based on what `foobar` features are being activated would be nice, but I don't think conditional compilation inside `foobar-lint` is how that should be accomplished.

**Reply by xFrednet** [reply]
> Wait, are you talking about the lint implementation itself using conditional compilation (so you could compile it with different feature sets before it ever runs) 

I'm talking about having conditional compilation in the lint implementation itself. It would be nice to also inspect `cfg` attributes, but that will not really be possible for now, there is at least no nice way for Clippy to do so. (See rust-lang/rust-clippy#6435 for example)

The main focus of this suggestion is to support lints for crates which use features to enable certain functionality. If this affects the lint message or maybe adds a check is up to the user IMO.

**Reply by jhpratt** [reply]
If you can inspect `cfg` attributes, that would inherent include `feature = "foo"`. We could provide a way to query the features enabled on any given crate.

I share the concern regarding this effectively limiting us from distributing precompiled lints, as was mentioned.

**Reply by flip1995** [reply]
Definitely +1 for this. Crates with different feature sets might want to ship different lints (or lint levels) depending on which features are enabled.

> I share the concern regarding this effectively limiting us from distributing precompiled lints, as was mentioned.

This is still a bad idea IMO (see https://github.com/rust-linting/design/issues/6#issuecomment-889703893). It just restricts us from adding nice features like the requested in the OP. There's a reason why cargo downloads and compiles source code and why crates.io doesn't store pre-compiled binaries.

**Reply by jhpratt** [reply]
Yeah, the outcome of this realistically depends on #6 as a whole.

**Reply by jhpratt** [reply]
I'm not sure this is something we should explicitly guarantee given the future possibility of pre-shipped dylibs that are sandboxed, which would necessarily preclude conditional compilation. Optional lint packages would go alongside whatever registry ends up being used.

**Reply by xFrednet** [reply]
Would it be okay for everyone to add this as an unstable feature for now?

I still believe that this would be very useful, and it should be quite easy to allow for with the currently planned discovery option for the start. Not allowing it would probably be more work. Therefore, I think a good solution would be to allow the usage of features for now but also document that this is unstable and might change in the future.

**Reply by jhpratt** [reply]
I mean I don't think our current plan for starting out would necessarily _restrict_ using cargo features, but it's not something I'd particularly like to encourage for the reasons stated. So as long as it's documented that the behavior isn't guaranteed for the future (similar to what I've documented for file system access, for example), that seems reasonable.

**Comment by xFrednet** [comment]
We should support and encourage the documentation of lints.

**Reply by jhpratt** [reply]
My thought was to have support for a `cargo lint --explain foo::bar` in addition to the messages already emitted.

**Reply by jhpratt** [reply]
^^ There could still be a command to generate docs for all lints present, using the provided explanation.

**Comment by jhpratt** [comment]
I just realized this was never explicitly stated anywhere: the linter should be able to work with IDEs. This likely means being able to output machine-readable JSON.

**Reply by xFrednet** [reply]
+1 I would make a stronger statement that we should use the diagnostic emission from rustc. That will automatically provide us with console & JSON output for diagnostics and ensures that the output structure is always the same (within the rust eco system).

**Reply by jhpratt** [reply]
I'd consider that an implementation detail, personally. Not that we shouldn't do it.
